### ISSUE TITLE
[v5] Service API PostgreSQL DB locks on SELECT queries 

### ISSUE THREAD

**Issue by ivan-sukhomlyn** [reportportal/reportportal#950]
**Describe the bug**

Huge **SELECT** SQL queries block itself. The number for queries slowly grows up and then fail down Service-API with the exception regarding JDBC connection pool size.
```
2020-05-06 04:46:22.852  WARN 1 --- [o-8585-exec-157] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 08006
2020-05-06 04:46:22.852 ERROR 1 --- [o-8585-exec-157] o.h.engine.jdbc.spi.SqlExceptionHelper   : An I/O error occurred while sending to the backend.
2020-05-06 04:46:22.852  WARN 1 --- [nio-8585-exec-6] com.zaxxer.hikari.pool.ProxyConnection   : HikariPool-1 - Connection org.postgresql.jdbc.PgConnection@4e57be9e marked as broken because of SQLSTATE(08006), ErrorCode(0)
org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
```
One time we had **> 500** SELECT queries in the waiting state for a tuple:

1-day active sessions graph from AWS RDS Performance Insights:

![Screenshot 2020-05-06 at 11 51 18](https://user-images.githubusercontent.com/55390366/81205385-49f6bd80-8fd3-11ea-95d3-2b246b08d8dc.png)

1-hour active sessions graph
![Screenshot 2020-05-06 at 14 44 34](https://user-images.githubusercontent.com/55390366/81205392-4bc08100-8fd3-11ea-9289-98725ecf5771.png)

SQL query:
```
select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_
```

PostgreSQL activity statistics with blocked queries:

```
postgres=> SELECT
postgres->     activity.pid,
postgres->     activity.usename,
postgres->     activity.query,
postgres->     blocking.pid AS blocking_id,
postgres->     blocking.query AS blocking_query
postgres-> FROM pg_stat_activity AS activity
postgres-> JOIN pg_stat_activity AS blocking ON blocking.pid = ANY(pg_blocking_pids(activity.pid));
  pid  | usename |                                                                                                                                                                                                                                                                                                  query                                                                                                                                                                                                                                                                                                  | blocking_id |                                                                                                                                                                                                                                                                                             blocking_query                                                                                                                                                                                                                                                                                              
-------+---------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 17042 | root    | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ |       27180 | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_
  1746 | root    | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ |       27180 | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_
```

PostgreSQL activity ordered by time duration:
```
postgres=> SELECT
pid,
now() - pg_stat_activity.query_start AS duration,
query,
state
FROM pg_stat_activity
WHERE (now() - pg_stat_activity.query_start) > interval '5 minutes' and state = 'active'
ORDER BY duration DESC NULLS LAST;
  pid  |    duration     |                                                                                                                                                                                                                                                                                                  query                                                                                                                                                                                                                                                                                                  | state  
-------+-----------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------
 24019 | 01:59:17.685468 | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 24081 | 01:56:16.740649 | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 20673 | 01:50:14.737422 | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 20702 | 01:26:06.933601 | select launch0_.id as id1_9_, launch0_.approximate_duration as approxim2_9_, launch0_.description as descript3_9_, launch0_.end_time as end_time4_9_, launch0_.has_retries as has_retr5_9_, launch0_.last_modified as last_mod6_9_, launch0_.mode as mode7_9_, launch0_.name as name8_9_, launch0_.number as number9_9_, launch0_.project_id as project10_9_, launch0_.rerun as rerun11_9_, launch0_.start_time as start_t12_9_, launch0_.status as status13_9_, launch0_.user_id as user_id14_9_, launch0_.uuid as uuid15_9_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
```

So we can see that queries that blocks another one can be in the active state for 2 hours in a row.

**Expected behavior**

**SELECT** SQL queries don't block other SELECT queries

**Additional context**
ReportPortal version - 5.0
PostgreSQL version - 10.10
DB instance - AWS RDS `rds.m5.xlarge` with 4 vCPU and 16GB of memory
DB size - 300 GB.

Why do you block `SELECT` SQL queries?
It seems like a bug. But maybe you have a reason to block similar requests.

**Comment by DzmitryHumianiuk** [comment]
@ivan-sukhomlyn they were addressed in 5.1 and 5.2
could you please give it a try
and double check if it still relevant or not?

**Comment by ivan-sukhomlyn** [comment]
@DzmitryHumianiuk Thanks for your reply! It's good to hear that this's already a known issue.
Unfortunately, we can't upgrade to the 5.1 version due to SAML integration doesn't work correctly in 5.1.
I will try to upgrade to the 5.2 version later when it will be available.
Do you have any ETA regarding the 5.2 version release?

**Comment by Yumfriez** [comment]
@ivan-sukhomlyn Could you please describe the flow which leads to this state?
Like what data (how much and for how long it's being reported) do you send to RP, what is happening with RP instance in parallel with this actions (like requests from the UI side for the launch that is being reported), etc.

**Comment by DzmitryHumianiuk** [comment]
>Do you have any ETA regarding the 5.2 version release?
May 15th

**Comment by ivan-sukhomlyn** [comment]
@DzmitryHumianiuk Thanks for info sharing.

@Yumfriez Mostly we process a lot of big launch logs (logs can be up to ~3000 strings). Autotests can be performed for 1-2 hours. We can have ~20-30 parallel launch execution at the same time for several projects. Also, we have about 100 active ReportPortal UI users.
I assume that blocks described later were led by several ReportPortal API restarts (due to GC overhead limit) when some heavy requests were in an active state. But I can't guarantee that this was a reason for this. Maybe somebody creates a lot of requests for launches and didn't wait for their results.

As for now, we don't have a lot of waiting tuple queries after I killed all blocked SQL queries.
But I was surprised that SELECT queries block other SELECT queries.

**Comment by evjlobanova** [comment]
@Yumfriez Could you please check the answer of Ivan?

**Comment by smore07** [comment]
I am also facing this same issue with 5.2 version


2020-06-02 17:55:51.147  WARN 6 --- [o-8585-exec-251] o.h.engine.jdbc.spi.SqlExceptionHelper   : SQL Error: 0, SQLState: 08006
api_1            | 2020-06-02 17:55:51.147 ERROR 6 --- [o-8585-exec-251] o.h.engine.jdbc.spi.SqlExceptionHelper   : An I/O error occurred while sending to the backend.
api_1            | 2020-06-02 17:55:51.147 ERROR 6 --- [o-8585-exec-251] o.s.t.i.TransactionInterceptor           : Application exception overridden by rollback exception
api_1            |

**Comment by evjlobanova** [comment]
We are working on this issue

**Comment by evjlobanova** [comment]
Fixed in the version 5.3

**Comment by ivan-sukhomlyn** [comment]
Thanks! Looking forward to a new release :)

**Comment by smore07** [comment]
Great news!!

**Comment by ivan-sukhomlyn** [comment]
Unfortunately, this bug reproduces at the 5.3 version as well with the PostgreSQL 12.3
API version - **5.3.0**

```
postgres=> SELECT  
    activity.pid,
    activity.usename,
    activity.query,
    blocking.pid AS blocking_id,
    blocking.query AS blocking_query
FROM pg_stat_activity AS activity
JOIN pg_stat_activity AS blocking ON blocking.pid = ANY(pg_blocking_pids(activity.pid));
  pid  | usename |                                                                                                                                                                                              
                                                                                                    query                                                                                                       
                                                                                                                                                                                           | blocking_id |      
                                                                                                                                                                                                                
                                                                       blocking_query                                                                                                                           
                                                                                                                                                                   
-------+---------+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
-------------------------------------------------------------------------------------------------------------------------------------------------------------------
 12738 | root    | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_retr5_8_, l
aunch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.start_ti
me as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ |       12515 | selec
t launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_retr5_8_, launch0_.last_modified as
 last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.start_time as start_t12_8_, laun
ch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.id=$1 for update of launch0_
 12742 | root    | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_retr5_8_, l
aunch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.start_ti
me as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ |       12515 | selec
t launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_retr5_8_, launch0_.last_modified as
 last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.start_time as start_t12_8_, laun
ch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.id=$1 for update of launch0_
 12743 | root    | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_retr5_8_, l
aunch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.start_ti
me as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ |       12515 | selec
t launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_retr5_8_, launch0_.last_modified as
 last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.start_time as start_t12_8_, laun
ch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.id=$1 for update of launch0_
....

(2486 rows)
```

**Comment by Yumfriez** [comment]
@ivan-sukhomlyn For how long does this query remain in blocking state?

**Comment by ivan-sukhomlyn** [comment]
Queries like these can be in blocked for more than 3-5 and longer
```
reportportal=>  SELECT  
  pid,           
  now() - pg_stat_activity.query_start AS duration,
  query,
  state                         
FROM pg_stat_activity
WHERE (now() - pg_stat_activity.query_start) > interval '3 minutes' and query LIKE '%select%';
  pid  |    duration     |                                                                                                                                                                                      
                                                                                                            query                                                                                               
                                                                                                                                                                                                   | state  
-------+-----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------
 25459 | 00:03:56.655151 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 25460 | 00:04:43.695612 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.id=$1 for update of launch0_   | active
 25780 | 00:04:25.201704 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 25786 | 00:05:02.25097  | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.id=$1 for update of launch0_   | active

```

**Comment by Yumfriez** [comment]
@ivan-sukhomlyn On our prod env during reporting we have these `select for update` statements too. If it doesn't block items' start/finish and these locks are removed after launch finish I suppose it's a normal reporting case. But if these locks remain even after all items were reported and launch was finished then we should investigate it again

**Comment by ivan-sukhomlyn** [comment]
We have a huge number of parallel launches at the same time that reports to several projects.
For example,  at this time, I have 197 queries that are in the blocked state and their execution duration are more than 30 minutes.

```
postgres=>  SELECT
  pid,
  now() - pg_stat_activity.query_start AS duration,
  query,
  state
FROM pg_stat_activity
WHERE (now() - pg_stat_activity.query_start) > interval '10 minutes' and query LIKE '%select%';
  pid  |    duration     |                                                                                                                                                                                      
                                                                                                            query                                                                                               
                                                                                                                                                                                                   | state  
-------+-----------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------
 17740 | 00:14:50.432814 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 18039 | 00:14:47.755076 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 14164 | 00:14:53.122569 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 18040 | 00:14:42.683775 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 18041 | 00:14:35.85019  | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 18045 | 00:14:09.004644 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 10885 | 00:15:38.518289 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 18056 | 00:11:52.875922 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 18046 | 00:13:38.772125 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 14166 | 00:14:44.010902 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 18047 | 00:13:36.175972 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
 14167 | 00:15:28.545135 | select launch0_.id as id1_8_, launch0_.approximate_duration as approxim2_8_, launch0_.description as descript3_8_, launch0_.end_time as end_time4_8_, launch0_.has_retries as has_ret
r5_8_, launch0_.last_modified as last_mod6_8_, launch0_.mode as mode7_8_, launch0_.name as name8_8_, launch0_.number as number9_8_, launch0_.project_id as project10_8_, launch0_.rerun as rerun11_8_, launch0_.
start_time as start_t12_8_, launch0_.status as status13_8_, launch0_.user_id as user_id14_8_, launch0_.uuid as uuid15_8_ from public.launch launch0_ where launch0_.uuid=$1 for update of launch0_ | active
...
(197 rows)
```
Our all of the 300 connections in the pool available for the API service are in the active state
![Screenshot 2020-09-25 at 12 53 35](https://user-images.githubusercontent.com/55390366/94254280-3d95e180-ff2f-11ea-8b08-6d916f83efb6.png)

**Comment by coufalja** [comment]
We've hit a similar issue, once we run a large number of parallel launches the RP UI is almost unusable. Tens of seconds wait times for simple API calls. By an investigation using a Profiler on our live instance, we have found out that it is somehow blocked by those select for update queries actually one in particular.
<img width="1014" alt="Screenshot 2020-09-29 at 12 02 38" src="https://user-images.githubusercontent.com/23357992/94655625-2ffda480-02ff-11eb-8f3e-8cdf57f22d6f.png">

Coming from this code
<img width="1207" alt="Screenshot 2020-09-29 at 12 02 48" src="https://user-images.githubusercontent.com/23357992/94655683-473c9200-02ff-11eb-8960-395ecfd35d60.png">

Comparison of all queries issued by RP
<img width="1203" alt="Screenshot 2020-09-29 at 12 02 16" src="https://user-images.githubusercontent.com/23357992/94655786-65a28d80-02ff-11eb-980a-da5847b4acb6.png">

**Comment by evjlobanova** [comment]
@coufalja which version of RP do you use?
                 does your instance have set up described here https://reportportal.io/docs/Optimal-Performance-Hardware?
                  If not, please add information about instance set up:  disck i/o, CPU, info about DB

cc @irynakozak2

**Comment by irynakozak2** [comment]
@ivan-sukhomlyn @coufalja that ticket has been reopened from our backlog. We will proceed with a reproducing and possible fix.

**Comment by coufalja** [comment]
Any news here? Would you mind sharing (briefly) what to check in the codebase? So we can attempt the fix ourselves and submit a PR. We have deadlocks in DB almost every other day and it is quite a PITA.

**Comment by coufalja** [comment]
I believe that the issue lies here https://github.com/reportportal/service-api/blob/5cadf3f9fad49b050461f080ba50dfdd7cf8054e/src/main/java/com/epam/ta/reportportal/core/item/impl/FinishTestItemHandlerImpl.java#L144 as the retrieve method obtains the Launch with the PessimisticLock for the duration of processing the test item. That is quite "inefficient" to say at least. Is the lock necessary to keep for the sake of simply updating the status in https://github.com/reportportal/service-api/blob/5cadf3f9fad49b050461f080ba50dfdd7cf8054e/src/main/java/com/epam/ta/reportportal/core/item/impl/status/ChangeStatusHandlerImpl.java#L114  Moreover the check will not pass in (blunt approximation ahead) 99% of the cases (It is not uncommon to have tens of thousands of test items in a single launch) resulting in no locking necessary in the first place.

**Comment by coufalja** [comment]
The profile after reversing the commit (and fixing tests) 
![Uploading Screenshot 2020-11-02 at 20.45.22.png…]()

**Comment by coufalja** [comment]
We are now running from the fork ^ and no Deadlocks since then, we even switched back for a while and got Deadlock within an hour.

**Comment by DzmitryHumianiuk** [comment]
@coufalja awesome!
We will take the for for performance testing in next sprint.

**Comment by ersuniljoshi** [comment]
Can some one help me how can I build a customer service api docker container out this branch(https://github.com/reportportal/service-api/pull/1319), we having been facing this issue since last 3 weeks and finding it really tough to keep the report portal up and running for long.

**Comment by coufalja** [comment]
@ersuniljoshi I already built the image for testing purposes here `fluffycroc/service-api:latest`. If you want to build it yourself then just checkout the branch and run Gradle `buildDocker` task.

**Comment by ersuniljoshi** [comment]
@coufalja Thanks for your response. We are currently on RP5.3.2.  Have you developed this on top of Rp5.3.2?

**Comment by coufalja** [comment]
It is based on 5.3.3 (I've changed the tag of the image accordingly to `fluffycroc/service-api:5.3.3`)

**Comment by coufalja** [comment]
@evjlobanova Any news here? I believe I have fixed both this issue and #1186 with https://github.com/reportportal/service-api/pull/1319 we are running the change for a month and no deadlock appeared since then.

**Comment by evjlobanova** [comment]
The task is in the current sprint. Actually, we cannot merge a sent pr, because it doesn't take into account all RP specific. 
But I hope we will fix this sprint. 
I will be updating this ticket accordinagly
